linkerd/app/src/dst/permit.rs

```diff
@@ -1,79 +1,96 @@
-    _error: PhantomData<fn(E)>,
 }
 
-// === impl PermitConfiguredDsts ===
+// === impl PermitProfile ===
 
-impl PermitConfiguredDsts {
-    pub fn new(
+impl PermitProfile {
+    pub(super) fn new(
         name_suffixes: impl IntoIterator<Item = Suffix>,
         nets: impl IntoIterator<Item = IpNet>,
     ) -> Self {
-        Self {
-            name_suffixes: Arc::new(name_suffixes.into_iter().collect()),
+        Self(Inner {
+            names: Arc::new(name_suffixes.into_iter().collect()),
             networks: Arc::new(nets.into_iter().collect()),
-            _error: PhantomData,
-        }
+        })
     }
+}
 
-    /// Configures the returned error type when the target is outside of the
-    /// configured set of destinations.
-    pub fn with_error<E>(self) -> PermitConfiguredDsts<E>
-    where
-        E: Into<Error> + From<Addr>,
-    {
-        PermitConfiguredDsts {
-            name_suffixes: self.name_suffixes,
-            networks: self.networks,
-            _error: PhantomData,
+impl<T> FilterRequest<T> for PermitProfile
+where
+    for<'t> &'t T: Into<Addr>,
+{
+    type Request = Addr;
+
+    fn filter(&self, t: T) -> Result<Addr, Error> {
+        let addr = (&t).into();
+        let permitted = self.0.matches(&addr);
+        tracing::debug!(permitted, "Profile");
+        if permitted {
+            Ok(addr)
+        } else {
+            Err(Rejected(()).into())
         }
     }
 }
 
-impl<E> Clone for PermitConfiguredDsts<E> {
-    fn clone(&self) -> Self {
-        Self {
-            name_suffixes: self.name_suffixes.clone(),
-            networks: self.networks.clone(),
-            _error: PhantomData,
-        }
+// === impl PermitResolve ===
+
+impl PermitResolve {
+    pub(super) fn new(
+        name_suffixes: impl IntoIterator<Item = Suffix>,
+        nets: impl IntoIterator<Item = IpNet>,
+    ) -> Self {
+        Self(Inner {
+            names: Arc::new(name_suffixes.into_iter().collect()),
+            networks: Arc::new(nets.into_iter().collect()),
+        })
     }
 }
 
-impl<T, E> FilterRequest<T> for PermitConfiguredDsts<E>
+impl<T> FilterRequest<T> for PermitResolve
 where
-    E: Into<Error> + From<Addr>,
-    for<'t> &'t T: Into<Addr>,
+    for<'t> &'t T: Into<Option<Addr>>,
 {
-    type Request = T;
+    type Request = Addr;
 
-    fn filter(&self, t: T) -> Result<T, Error> {
-        let addr = (&t).into();
-        let permitted = match addr {
-            Addr::Name(ref name) => self
-                .name_suffixes
-                .iter()
-                .any(|suffix| suffix.contains(name.name())),
+    fn filter(&self, t: T) -> Result<Addr, Error> {
+        let permitted = (&t).into().and_then(|addr| {
+            if self.0.matches(&addr) {
+                Some(addr)
+            } else {
+                None
+            }
+        });
+        tracing::debug!(permitted = %permitted.is_some(), "Resolve");
+        permitted.ok_or_else(|| Rejected(()).into())
```

Comment:
i was a little surprised at first that there are two separate implementations for resolve and profiles which _look_ equivalent after a quick skim, but i didn't notice the `Option` until i took a second look at it. it does occur to me that we could probably only have one permit implementation if we added `Into<Option<Addr>>` impls that always return `Some` for profile targets. 

i don't think this is like, important --- the small amount of duplication is totally fine with me, i was just surprised until i looked closer.

Reply:
I had something like that initially but i think it's a bit clearer to have them separated (for now at least).